Observer pods: docs (#287)
openshift/ci-docs: content/en/docs/architecture/ci-operator-internals/observer-pods.md

---
title: "Observer Pods"

> description: A description of how observer pods work

Reviewer: possibly without **A**

---

## Motivation

> Observer pods have been requested and designed to make the life easier for an e2e test author who wants to:

Reviewer: ~~requested~~

Reviewer: make life easier (without the?)

> - have an active non-intruse 'observing agent' throughout the entire exection of a(n e2e) test

Reviewer: exection->execution

> - gather information from a(n ephemeral) cluster

Reviewer: (including ephemeral instances)

- get the previous functionalities without having to make changes to every single step of the e2e test

> Observer pods are called this way because they are supposed just to watch/observer without affecting the exection of the current test.  

Reviewer: exection-> execution and they are supposed to observer

> If an observer pod fails, the failure will still be visible inside logs but the test won't ever be affected.

Reviewer: i think we can cross ~~ever~~

There are different ways of defining observers and including/excluding them in a multi-step test workflow:

### Embedded in a test definition
{{< highlight yaml >}}
tests:
- as: e2e
  steps:
    test:
    - as: first-step
      commands: echo 'this is a test'
      ...
      observers:
      - name: my-awesome-observer
        # Only one of the following two should be specified
        from: cli # This image will be used to run the observer
        # Literal image stream tag that will be used to run the observer
        from_image:
          ...
        # Commands that are going to be executed
        commands: echo 'this is an observer' 
        # Regular resources constraint definition for a pod running this observer
        resources: 
          ...
{{< / highlight >}}

### Referring to existing observers
It is possible to reference observers that already exist:
{{< highlight yaml >}}
tests:
- as: e2e
  steps:
    pre:
      ...
    test:
      ...
    post:
      ...
    observers:
      enable:
      # Observers 'observer-1' and 'observer-2' have to exist somewhere 
      # inside the step registry
      - observer-1
      - observer-2
{{< / highlight >}}

> Observers definitions are additive, that means the final set of observers will be the union between any

Reviewer: Observers -> Observers'

Reviewer: that means -> which means

> observer defined inside test step definition and the ones into `observers` stanza.

Reviewer: and the ones **defined in** the...?
ones into -> ones defined in the

Assuming the following configuration:
{{< highlight yaml >}}
tests:
- as: e2e
  steps:
    test:
    - as: first-step
      commands: echo 'this is a test'
      ...
      observers:
      - observer-3
    observers:
      enable:
      - observer-1
      - observer-2
{{< / highlight >}}

The final set of observers will be: `{ observer-3, observer-1, observer-2 }`.

> The position matters: observers are added to the set as they are encountered, from the top to the bottom of the configuration.

Reviewer: position-> ordering

### Disable a predefined observer
The `observers` stanza found inside `steps` is made up in this way:
{{< highlight yaml >}}
observers:
  # List of observers that will be included in the test execution
  enable:
  - observer-1
  - observer-2
  # List of observers that *won't* be included in the test execution
  disable:
  - observer-3
{{< / highlight >}}

The `disable` stanza let the user selectively decide what observers not to include.
Assuming the following configuration:
{{< highlight yaml >}}
tests:
- as: e2e
  steps:
    test:
    - as: first-step
      commands: echo 'this is a test'
      ...
      observers:
      - observer-3
    observers:
      enable:
      - observer-1
      - observer-2
      disable:
      - observer-3
{{< / highlight >}}
The final set of observers will be: `{ observer-1, observer-2 }`.

> ### From scratch on Step Registry

Reviewer: From scratch on Step Registry -> Writing from scratch inside Step Registry?

Writing an observer as a step inside the registry looks almost like a regular test.

Observer's definition file must have `-observer.yaml` suffix. This is how the configuration looks like:
{{< highlight yaml >}}
observer:
  name: observer
  # Only one of the following two should be specified
  from: os # This image will be used to run the observer
  # Literal image stream tag that will be used to run the observer
  from_image:
    ...
  # This file has to be inside the same observer folder
  commands: observer-commands.sh
  resources:
    requests:
      cpu: 10m
      memory: 10Mi
  documentation: |-
    A simple observer
{{< / highlight >}}

> As it is for regular tests, observer's payload file must have `-commands` suffix (file extension has not

Reviewer: **the** observer's payload?

been taken into account this time): 
{{< highlight bash >}}
#!/bin/bash

set -o errexit
set -o nounset
set -o pipefail
set -o xtrace

echo "I'm going to silently observe your test"
{{< / highlight >}}

The final folder content wuold be, in this case:
- `observer-observer.yaml`
- `observer-commands.sh`

## Execution Model
Observers are started before any test from the `pre:` chain does and they keep staying alive till
the last step from the `test:` chain completes.  

When an observer fails, its failure will be reported inside logs and jUnit files but the test
will keep going without being affected at all.  

When the last test from `test:` chain finishes, a `SIGTERM` is sent to every observer and its related pod is deleted. After the test completes, no 'observer pods' pods will be found inside `ci-op-xxxx` namespace.

Artifacts are being collected as they were regular tests.

It's a good habit writing observer payload following this approach:
{{< highlight bash >}}
#!/bin/bash

set -o errexit
set -o nounset
set -o pipefail
set -o xtrace

function cleanup() {

>     echo 'this will be executed last'

Reviewer: last-> at the end?

}

trap cleanup EXIT

# Do something here

> # An observer could be killed in the middle of its execution!

Reviewer: killed or sigterm is sent?

Author: Kubernetes try to send a `SIGTERM` signal firstly and then, after a grace period of n seconds, it `SIGKILL`s the container.

...
{{< / highlight >}}
## Features
### Shared directory

> `$SHARED_DIR` is still being monted on observer pods but it is *read-only*.

Reviewer: monted -> mounted

> As regular tests modify it, kubernetes does the heavy lift of propagating the changes.

Reviewer: kubernetes -> Kubernetes

> The following example shows what to expect from the observer point of view when a test write something on

Reviewer: The following example shows what to expect from the observer**'s** point of view when a test **writes** something on
the `$SHARED_DIR`.

the `$SHARED_DIR`.  

Timeline of events:
- Observer `observer-1` starts, `$SHARED_DIR` is empty
- Test `test-1` starts
- Test `test-1` writes file `dummy.txt` inside `$SHARED_DIR`
- Test `test-1` ends gracefully

> - After a variable amount of time (up to ~2 minutes) kubernetes updates `$SHARED_DIR` content on `observer-1`

Reviewer: kubernetes->Kubernetes

> - At this point only observer pod `observer-1` sees `dummy.txt`

Reviewer: At this point-> At this point,

> ### Kubeconfig

Reviewer: Kubeconfig -> `kubeconfig`?

> Observers will get a `read-write` kubeconfig as soon as it is provided by some tests or any other agent

Reviewer: kubeconfig copy?

at the path specified in `$KUBECONFIG`.

> It is safe to check whether it exist or not doing so:

Reviewer: It is safe to check whether it exists or not by invoking the following code:

{{< highlight bash >}}
if [ -f "$KUBECONFIG" ]; then
    ...
fi
{{< / highlight >}}

> Kubeconfig is treated differently from `$SHARED_DIR`, so observers could receive at any time. Please do not make any assumption about order of events.

Reviewer: `kubeconfig` is treated differently from `$SHARED_DIR`, so observers could receive **it** at any time. Please do not make any **assumptions** about **the** order of events.

---
title: "Observer Pods"
description: Description of how observer pods work
---

## Motivation
Observer pods have been requested and designed to make life easier for an e2e test author who wants to:

> - have an active non-intruse 'observing agent' throughout the entire execution of a(n e2e) test

Reviewer 2: ```suggestion
- have an active non-intrusive 'observing agent' throughout the entire execution of an e2e test
```

- gather information from a cluster (including ephemeral instances)
- get the previous functionalities without having to make changes to every single step of the e2e test

> Observer pods are called this way because they are supposed to observer without affecting the execution of the current test.  

Reviewer 2: ```suggestion
Observer pods are named as such because they are supposed to observe without affecting the execution of the current test.  
```

> If an observer pod fails, the failure will still be visible inside logs but the test won't be affected.

Reviewer 2: ```suggestion
If an observer pod fails, the failure will still be visible inside logs, but the test won't be affected.
```

        resources: 
          ...
{{< / highlight >}}

### Referring to existing observers
It is possible to reference observers that already exist:
{{< highlight yaml >}}
tests:
- as: e2e
  steps:
    pre:
      ...
    test:
      ...
    post:
      ...
    observers:
      enable:
      # Observers 'observer-1' and 'observer-2' have to exist somewhere 
      # inside the step registry
      - observer-1
      - observer-2
{{< / highlight >}}

Observers' definitions are additive, which means the final set of observers will be the union between any
observer defined inside test step definition and the ones defined in the `observers` stanza.

Assuming the following configuration:
{{< highlight yaml >}}
tests:
- as: e2e
  steps:
    test:
    - as: first-step
      commands: echo 'this is a test'
      ...
      observers:
      - observer-3
    observers:
      enable:
      - observer-1
      - observer-2
{{< / highlight >}}

The final set of observers will be: `{ observer-3, observer-1, observer-2 }`.
The ordering matters: observers are added to the set as they are encountered, from the top to the bottom of the configuration.

### Disable a predefined observer

> The `observers` stanza found inside `steps` is made up in this way:

Reviewer 2: ```suggestion
The `observers` stanza found inside `steps` is as follows:
```

{{< highlight yaml >}}
observers:
  # List of observers that will be included in the test execution
  enable:
  - observer-1
  - observer-2
  # List of observers that *won't* be included in the test execution
  disable:
  - observer-3
{{< / highlight >}}

> The `disable` stanza let the user selectively decide what observers not to include.

Reviewer 2: ```suggestion
The `disable` stanza lets the user selectively decide which observers to exclude.
```

Assuming the following configuration:
{{< highlight yaml >}}
tests:
- as: e2e
  steps:
    test:
    - as: first-step
      commands: echo 'this is a test'
      ...
      observers:
      - observer-3
    observers:
      enable:
      - observer-1
      - observer-2
      disable:
      - observer-3
{{< / highlight >}}
The final set of observers will be: `{ observer-1, observer-2 }`.

### Writing from scratch inside Step Registry
Writing an observer as a step inside the registry looks almost like a regular test.

> Observer's definition file must have `-observer.yaml` suffix. This is how the configuration looks like:

Reviewer 2: ```suggestion
The observer's definition file must have `-observer.yaml` suffix. The configuration looks like:
```

{{< highlight yaml >}}
observer:
  name: observer
  # Only one of the following two should be specified
  from: os # This image will be used to run the observer
  # Literal image stream tag that will be used to run the observer
  from_image:
    ...
  # This file has to be inside the same observer folder
  commands: observer-commands.sh
  resources:
    requests:
      cpu: 10m
      memory: 10Mi
  documentation: |-
    A simple observer
{{< / highlight >}}

As it is for regular tests, the observer's payload file must have `-commands` suffix (file extension has not
been taken into account this time): 
{{< highlight bash >}}
#!/bin/bash

set -o errexit
set -o nounset
set -o pipefail
set -o xtrace

echo "I'm going to silently observe your test"
{{< / highlight >}}

> The final folder content wuold be, in this case:

Reviewer 2: ```suggestion
The final folder content would be, in this case:
```

- `observer-observer.yaml`
- `observer-commands.sh`

## Execution Model

> Observers are started before any test from the `pre:` chain does and they keep staying alive till

Reviewer 2: ```suggestion
Observers are started before any test from the `pre:` chain starts and they stay alive until
```

the last step from the `test:` chain completes.  

> When an observer fails, its failure will be reported inside logs and jUnit files but the test

Reviewer 2: ```suggestion
When an observer fails, its failure will be reported inside logs and jUnit files, but the test
```

will keep going without being affected at all.  

> When the last test from `test:` chain finishes, observers are going to be deleted by Kubernetes: see [here](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination) to check more in detail how pods termination works. After the test completes, no 'observer pods' pods will be found inside `ci-op-xxxx` namespace.

Reviewer 2: ```suggestion
When the last test from `test:` chain finishes, observers are going to be deleted by Kubernetes: see [here](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination) for more details on how pod termination works. After the test completes, no 'observer pods' pods will be found inside `ci-op-xxxx` namespace.
```

> Artifacts are being collected as they were regular tests.

Reviewer 2: ```suggestion
Artifacts are collected as if they were regular tests.
```

> It's a good habit writing observer payload following this approach:

Reviewer 2: ```suggestion
It's a good habit to write the observer payload following this approach:
```

{{< highlight bash >}}
#!/bin/bash

set -o errexit
set -o nounset
set -o pipefail
set -o xtrace

function cleanup() {
    echo 'this will be executed at the end'
}

trap cleanup EXIT

# Do something here

# An observer could be killed in the middle of its execution:
# firstly, Kubernetes sends it a SIGTERM and then, after a grace period of n seconds, 
# it just kills (SIGKILL) it.
...
{{< / highlight >}}
## Features
### `$SHARED_DIR`

> `$SHARED_DIR` is still being mounted on observer pods but it is *read-only*.

Reviewer 2: ```suggestion
`$SHARED_DIR` is still mounted on observer pods but it is *read-only*.
```

As regular tests modify it, Kubernetes does the heavy lift of propagating the changes.

The following example shows what to expect from the observer's point of view when a test writes something on
the `$SHARED_DIR`.  

Timeline of events:
- Observer `observer-1` starts, `$SHARED_DIR` is empty
- Test `test-1` starts
- Test `test-1` writes file `dummy.txt` inside `$SHARED_DIR`
- Test `test-1` ends gracefully
- After a variable amount of time (up to ~2 minutes) Kubernetes updates `$SHARED_DIR` content on `observer-1`
- At this point, the observer pod `observer-1` sees `dummy.txt`

### `$KUBECONFIG`
Observers will get a `read-write` kubeconfig copy as soon as it is provided by some tests or any other agent
at the path specified in `$KUBECONFIG`.

> It is safe to check whether it exist or not by invoking the following code:

Reviewer 2: ```suggestion
It is safe to check for its existence by invoking the following code:
```
